proposals/source-install.md

```diff
@@ -0,0 +1,1133 @@
+   copies all Python libraries from the Python interpreter's site package plus
+   any scripts (e.g., `pyinstaller`) into the virtual environment being used
+   for the build.
+   
+3. Run `pip install` directly on the AWS CLI v2 codebase to do an offline,
+   in-tree build and install of the AWS CLI v2 into the build virtual
+   environment. This is done by including the follow pip flags:
+   * [`--no-build-isolation`](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-no-build-isolation)
+   * [`--use-feature=in-tree-build`](https://pip.pypa.io/en/stable/cli/pip_install/#local-project-installs)
+   * [`--no-cache-dir`](https://pip.pypa.io/en/stable/cli/pip_install/#caching)
+   * [`--no-index`](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-no-index)
+   
+4. If the `--install-type` was set to `portable-exe` in the `configure`
+   command, run [`pyinstaller`](https://www.pyinstaller.org/) to build a
+   standalone executable.
+   
+
+#### Install steps
+
+When running `make install`, the following steps are run:
+
+1. Move the built virtual environment (if the install type is `system-sandbox`)
+   or standalone executable (if the install type is a `portable-exe`) to the
+   configured install directory (i.e., `<libdir>/awscli`).
+   
+2. Create symlinks for both the `aws` and `aws_completer` in the configured
+   bin directory (i.e., at `<bindir>/aws` and `<bindir>/aws_completer`).
+
+
+### Provisional support for PEP 517
+
+[PEP 517](https://www.python.org/dev/peps/pep-0517/) specifies a build-system
+independent format for building Python packages. The AWS CLI v2 code base
+is PEP 517 compliant in that a PEP 517 compliant build tool can build sdists
+and wheels of the AWS CLI v2, which in turn can be used to install the
+AWS CLI v2 onto a system. For example, users can build a wheel and install it
+through pip:
+```
+pip wheel . --wheel-dir dist/
+pip install dist/awscli-2.2.1-py3-none-any.whl
+```
+
+#### Meaning of "provisional support"
+
+Provisional support means that PEP 517 compliance is maintained as a best
+effort, and there is no guarantee that compliance will be maintained in the
+future. Users should only directly rely on this functionality as a last resort
+because its support is contingent on the implementation detail that the AWS
+CLI is written in Python.
+
+
+#### Required code base changes for PEP 517 support
+
+To make the AWS CLI v2 codebase compliant, it requires the following changes:
+
+* Port all information from the `setup.py` to the `setup.cfg` that can be
+  programmatically parsed for runtime dependencies.
+  
+* Introduce a `pyproject.toml` file (see [PEP 518](https://www.python.org/dev/peps/pep-0518/))
+  and in-tree build backend that builds the autocomplete index when building
+  both the sdist and wheel
+  (see [PEP 517](https://www.python.org/dev/peps/pep-0517/#in-tree-build-backends)).
+  
+* Pull in and maintain the unreleased `botocore` `v2` branch along with
```

Comment:
This makes me so sad. Just release `botocore` v2 and `s3transfer`! So much easier for the entire community.

Reply:
> So much easier for the entire community.

Could you elaborate on what specifically will make things easier? If we were to release a new major version of botocore using what is currently on the v2 branch, it would have the following implications:

* Right now the changes on the v2 branch is only scoped to [changing defaults and removing functionality](https://github.com/boto/botocore/pulls?q=is%3Apr+is%3Aclosed+base%3Av2+NOT+develop+in%3Atitle+NOT+bump+in%3Atitle+). The major version release would not address many of the popular feature requests that would only be feasible in a major version bump (e.g., [Async support](https://github.com/boto/botocore/issues/458), [modularization](https://github.com/boto/botocore/issues/1543), [out of the box typing support](https://github.com/boto/boto3/issues/1055)) and likely require larger rewrites of botocore (or even replace botocore outright). Currently, the maintainers of the Python SDK are exploring how to go about making these changes such as in their work writing the [Amazon Transcribe Streaming SDK](https://github.com/awslabs/amazon-transcribe-streaming-sdk).

* It would require a major version bump of `boto3` and force many downstream Python packages to account for new major versions of `botocore`/`boto3` (e.g., by updating their version range on `botocore`/`boto3` or making major version bumps themselves). If appropriate adjustments aren't made, then these downstream packages would not be usable alongside the latest version of `botocore`.

In the end, given the current state of changes in the `v2` branch, a major version bump does not seem worth the churn it would cause to the community, especially because it would be missing key features that would require another major version bump down the road.

Reply:
That makes sense. I didn't realize the botocore v2 dependency of the released v2 awscli was so incomplete! That's a bit of an ugly dependency tree for the v2 cli. Sorry I'm so bitter. Appreciate the detailed reply. I'll try to be more kind.

So, would it not work to publish the dev releases of botocore v2? There's no compatibility promises that way (except to yourselves). And you could still include "normal" python dependency references and installation that way. Pip won't install dev releases unless you tell it to explicitly.

* https://www.python.org/dev/peps/pep-0440/#developmental-releases
* https://pip.pypa.io/en/stable/cli/pip_install/#pre-release-versions

Reply:
> Sorry I'm so bitter. Appreciate the detailed reply. I'll try to be more kind.

No worries. The frustration is understandable given our inconsistent communication with the community about the conversations/decisions happening around issue prioritization and project direction (which we are trying to improve).

> I didn't realize the botocore v2 dependency of the released v2 awscli was so incomplete! That's a bit of an ugly dependency tree for the v2 cli.

I completely agree. Originally the intention was to fast-follow with a major version bump of botocore/boto3 so we would not be sitting on this unreleased branch of botocore for long, but the direction has changed as we realized it was going to require much larger breaking changes to implement the larger features that the community wants.

> So, would it not work to publish the dev releases of botocore v2?

These are the main concerns:

* If you are going to try to install the AWS CLI v2 side-by-side with `boto3` as a Python package installation, it would not be feasible because `boto3` relies on v1 of `botocore` and the CLI v2 would require this dev version. For Python developers, they should be savvy enough to install the CLI into a separate `virtualenv` or use `pipx` . However, this becomes really problematic for Linux distro maintainers where they only package Python packages at the system level without a virtual environment (because virtualenvs work against their packaging principles). As alternative, distro maintainers should be able package using the `portable-exe` install type, but if we were to rely on a public dev version for `botocore` v2, it would block their ability to treat packaging the AWS CLI v2 as a typical Python package. And we want to make sure we are minimizing roadblocks for distro maintainers to pull in v2.

* Because a larger rewrite will be needed for the larger changes that the Python SDK maintainers want to make, there is a reasonable chance it warrants creating a new package name (e.g. `botocore2` or even depart from the `botocore` name entirely). If we start releasing public dev versions of `botocore` v2 and decide to use a new package name for v2, it creates an awkward story where we've been releasing dev versions of `botocore` v2 but the actual GA version is a completely separate package on PyPI.

In the end, I do not like that we'd be pulling in the unreleased botocore branch into the codebase, but at the current moment, seems like the most attainable option that unblocks issues installing the AWS CLI v2 from source and is also a two-way door. Long term, I would like to see us move a way from this approach through any of these options below:

* Integrate with whatever the new major version of the Python SDK would be.
* Figure out if there is a path forward to re-integrate with `botocore` v1 if the new major version of the Python SDK is not a viable path forward. Theoretically it seems feasible? But there may be some issues that make it not straightforward. 
* Rely on another language SDK or more directly rely on the [Python bindings for the AWS Common Runtime](https://github.com/awslabs/aws-crt-python) as that project expands.

Thoughts?

Reply:
Ahh, so if a botocore v2.dev version is installed in the `botocore` namespace, I can see how that could conflict with other python development. Practices in this area of python have [come a long way the last few years](https://ben11kehoe.medium.com/my-python-setup-77c57a2fc4b6), and not using a virtualenv for that python development, and not using `pipx` for cli installs...yeah that's going to end in pain.

In that case, a new namespace, e.g. `botocore2`, and publishing _only_ dev versions for now makes a lot of sense to me. It would also allow other python libraries to start integrating with it without jumping through hoops, to explore new features or any interface changes they may need to make.

I don't think I have enough experience with the distro packaging world to understand this piece:

>if we were to rely on a public dev version for botocore v2, it would block [distro maintainers] ability to treat packaging the AWS CLI v2 as a typical Python package

Why wouldn't they be able to use the public dev version? I'm probably off base, but I'd think they would also package "botocore2" and establish the dependency the same way they would for other packages?

Reply:
> I don't think I have enough experience with the distro packaging world to understand this piece:

@lorengordon I haven't followed the `botocore` situation that closely, but here's what's going on in general terms.

If `botocore` v2 were packaged in a separate namespace, then packaging a public dev version would be easy - you just install both next to each other.  But that's not what's happening: at least for now, they're continuing to use the `botocore` namespace.

If packagers ship it as-is, as `botocore`, the backwards incompatibilities risk impacting a huge number of apps.  If packagers patch it to `botocore2`, that risk is resolved.  But then every app that wants v2 needs to be patched as well.  So it's not very attractive.

Debian has a transition process to support incompatible upgrades - new version is uploaded into experimental, apps that depend on it are tested, bugs are filed for upgrades, when they're all fixed the library can be upgraded in unstable.  But until v2 has a stable api and version, this is unattractive: if another update causes more compatibility breaks, another transition would be required for the update.

Reply:
> Thoughts?

@kyleknap I agree that bundling the `botocore` source isn't ideal, but I agree that it sounds like a reasonable trade-off.  Hopefully the situation eventually gets simpler, and the embedded copy can go away.

Also, thanks for thinking about the challenges facing distros, it's definitely appreciated.

Reply:
> Also, thanks for thinking about the challenges facing distros, it's definitely appreciated.

@rvandegrift Yep! Removing these challenges is important to us. While our official installers (PKG, MSI, Linux installer) are not going away, long term it is going to be a better experience for everyone if they can install the AWS CLI v2 through their standard package manager.

Reply:
> In that case, a new namespace, e.g. `botocore2`, and publishing only dev versions for now makes a lot of sense to me. It would also allow other python libraries to start integrating with it without jumping through hoops, to explore new features or any interface changes they may need to make.

@lorengordon Yeah I think is the strategy that we would like to see if we went down the route of integrating with the new major version of the Python SDK. Specifically, the maintainers of the AWS Python SDK would publish `0.x.y` versions under the new namespace and the AWS CLI would try to keep up with any interface changes they made while progress was still being made. This approach would also allow other Python developers to give feedback on the changes being made.

If we were to try to integrate with the package before it was GA'd, we'd definitely need to be mindful of a couple things as a GA'd project relying on a non-GA'd project:

* We should make sure to integrate once we are confident that there would be no significant changes to the client/session interfaces. If there were significant changes to those interfaces, the hope would be that there is some back compat interface to ease the migration as it could be potentially an enormous effort to update all client usage and we would only have to make the change once.
 
* The moment we do integrate it makes it more difficult the new project to make larger changes quickly as a large GA'd project is relying on it. So if we just put whatever was on the `v2` branch published under a`botocore2` namespace and immediately started relying on it, it will be impede/slow down progress on many of the broader, more invasive changes the maintainers of the Python SDK team do want to make. This was similar to the situation when AWS CLI v1 GA'd but `botocore` was still a non-GA'd project. While it was still feasible to make larger changes in the `botocore` codebase, making changes was not straightforward, and at times, design choices were made with emphasis on the AWS CLI usage, resulting in tight coupling and interfaces that could be better generalized/defined. With this new major version, it would be ideal to see the larger changes have a stronger foundation before on-boarding the AWS CLI to not too strongly influence the approach.

Reply:
>The moment we do integrate it makes it more difficult the new project to make larger changes quickly as a large GA'd project is relying on it. So if we just put whatever was on the v2 branch published under abotocore2 namespace and immediately started relying on it, it will be impede/slow down progress on many of the broader, more invasive changes the maintainers of the Python SDK team do want to make.

I'm not sure I understand how this is true. It is already the case that you [pin to a specific botocore v2 tag](https://github.com/aws/aws-cli/blob/258dc7c50243488ba7ee0d8f8487c5bcaecc9ff1/setup.py#L27), and increment that for each release, testing as you go, right? So, already a large GA'd project is relying on the botocore v2 implementation. Are you saying the botocore project _wants_ to start v2 over again, basically? Or that you want to relax that strict pin in awscli v2?